### PR TITLE
feat(security): integrate LeakDetector into output response pipeline

### LINKED ISSUE #1979: [Feature]: Integrate LeakDetector + PromptGuard into response pipeline (output guardrail)
### Problem

ZeroClaw has two well-implemented security modules — **LeakDetector** (384 LOC, credential pattern scanning) and **PromptGuard** (360 LOC, injection detection) — but neither is called in the production response pipeline. Both are exported, tested internally, but have **zero production invocations**.

This means:
- LLM responses containing leaked API keys, JWT tokens, or database URLs are sent directly to users via channels
- Inbound user messages bypass prompt injection detection entirely

The three-layer guardrail model (input → tool → **output**) recommended by [Meta AI](https://ai.meta.com/blog/practical-ai-agent-security/), [OpenAI Agents SDK](https://openai.github.io/openai-agents-python/guardrails/), and [NVIDIA](https://developer.nvidia.com/blog/practical-security-guidance-for-sandboxing-agentic-workflows-and-managing-execution-risk/) is partially built but not wired in.

### Current State (source-verified)

| Module | File | Lines | Status |
|--------|------|-------|--------|
| LeakDetector | `src/security/leak_detector.rs` | 384 | ✅ Implemented, ❌ never called |
| PromptGuard | `src/security/prompt_guard.rs` | 360 | ✅ Implemented, ❌ never called |
| Tool sanitization | `src/channels/mod.rs` | `sanitize_channel_response()` | ✅ Active (strips `<tool_call>` tags + JSON artifacts) |

### Existing interception points

**Gateway flow** (WhatsApp, Slack, Discord, etc.):
```
LLM response → sanitize_gateway_response() → channel.send()
                    ↑ only strips tool artifacts
```

**CLI flow** (`src/agent/loop_.rs:1594`):
```
LLM response → channel.send()   ← no sanitization at all
```

### Proposal

**Phase 1 (minimal, low risk):** Wire LeakDetector into the existing `sanitize_channel_response()` function in `src/channels/mod.rs`. This is a single-point change that covers all channels.

```rust
// In sanitize_channel_response(), after tool artifact stripping:
let leak_detector = LeakDetector::new();
match leak_detector.scan(&sanitized) {
    LeakResult::Clean => sanitized,
    LeakResult::Detected { redacted, .. } => {
        tracing::warn!("Credential leak redacted before channel send");
        redacted
    }
}
```

**Phase 2 (optional):** Wire PromptGuard on inbound messages at a centralized point (gateway webhook handlers or a shared middleware).

**Phase 3 (optional):** Add configurable output guardrail trait for user-defined validators (content policy, topic scope, format compliance).

### Why this matters

- **#1805** and **#1811** are strengthening input-side detection (Aho-Corasick, Perplexity Filtering). Output-side is the natural complement.
- LeakDetector already covers: Stripe keys, OpenAI/Anthropic/Google API keys, GitHub tokens, AWS credentials, private keys, JWTs, database URLs.
- The code is **already written and tested** — this is an integration task, not a feature build.

### References

- Meta "Rule of Two": https://ai.meta.com/blog/practical-ai-agent-security/
- OpenAI Guardrails SDK: https://openai.github.io/openai-agents-python/guardrails/
- NVIDIA sandbox guidance: https://developer.nvidia.com/blog/practical-security-guidance-for-sandboxing-agentic-workflows-and-managing-execution-risk/
- NeMo Guardrails: https://github.com/NVIDIA/NeMo-Guardrails
- AWS Security Scoping Matrix: https://aws.amazon.com/blogs/security/the-agentic-ai-security-scoping-matrix-a-framework-for-securing-autonomous-ai-systems/

### PR DESCRIPTION
**Pull request by Mike-7777777**
## Summary

Describe this PR in 2-5 bullets:

- Base branch target (`main`): `main`
- Problem: ZeroClaw's `LeakDetector` (384 LOC, regex-based credential scanner) exists in `security/` but was never wired into production response pipelines, leaving all 15+ channel adapters, gateway webhooks, WebSocket handlers, and CLI output paths vulnerable to credential leakage through LLM responses.
- Why it matters: Without output guardrails, any API key, JWT, PEM private key, or database URL that appears in an LLM response or tool execution result is sent to end users verbatim -- a critical security gap in the three-layer guardrail model (input - tool - output).
- What changed: Integrated `LeakDetector` into all four output data paths via `sanitize_channel_response()` (channels/gateway/WS) and `apply_output_guardrail()` (CLI), added `[security.output_guardrail]` config section with three action modes (redact/warn/block), added `leak_sensitivity` bounds validation, and fixed WS fallback path that returned raw tool output.
- What did **not** change (scope boundary): `LeakDetector` implementation itself is unchanged; draft streaming sanitization and PromptGuard input-side integration are deferred to Phase 2; no new dependencies introduced.

## Label Snapshot (required)

- Risk label (`risk: low|medium|high`): `risk: high`
- Size label (`size: XS|S|M|L|XL`, auto-managed/read-only): `size: M`
- Scope labels (`core|agent|channel|config|cron|daemon|doctor|gateway|health|heartbeat|integration|memory|observability|onboard|provider|runtime|security|service|skillforge|skills|tool|tunnel|docs|dependencies|ci|tests|scripts|dev`, comma-separated): `security, config, channel, gateway, agent`
- Module labels (`<module>: <component>`, for example `channel: telegram`, `provider: kimi`, `tool: shell`): `security: core`, `config: core`, `gateway: ws`, `agent: loop`
- Contributor tier label (`trusted contributor|experienced contributor|principal contributor|distinguished contributor`, auto-managed/read-only; author merged PRs >=5/10/20/50): N/A (auto-managed)
- If any auto-label is incorrect, note requested correction: None

## Change Metadata

- Change type (`bug|feature|refactor|docs|security|chore`): `security`
- Primary scope (`runtime|provider|channel|memory|security|ci|docs|multi`): `security`

## Linked Issue

- Closes #1979
- Related #1805, #1811 (input-side hardening)
- Depends on # N/A
- Supersedes # N/A
- External tracking link(s) (optional): N/A

## Validation Evidence (required)

Commands and result summary:

```bash
cargo fmt --all -- --check
cargo clippy --all-targets -- -D warnings
cargo test
```

- Evidence provided (test/log/trace/screenshot/perf): `cargo fmt` passes clean. `cargo check` produces zero new errors -- all build errors are pre-existing upstream `Cargo.lock` desync (upstream `main` CI also fails identically). 18 new unit tests added and verified locally via `cargo test` on isolated modules.
- If any command is intentionally skipped, explain why: `cargo clippy` and full `cargo test --locked` are blocked by pre-existing upstream `Cargo.lock` desync that also fails on `main` branch CI. Individual test modules verified via `cargo test --lib channels::tests` and `cargo test --lib config::tests`.

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`): No
- New external network calls? (`Yes/No`): No
- Secrets/tokens handling changed? (`Yes/No`): Yes
- File system access scope changed? (`Yes/No`): No
- If any `Yes`, describe risk and mitigation: Secrets handling improved -- `LeakDetector` now scans all outbound LLM responses for credential patterns (Stripe/OpenAI/Anthropic/Google/GitHub API keys, AWS credentials, JWTs, PEM keys, DB URLs) and redacts them before reaching any output channel. Risk: false positives on strings resembling API keys. Mitigation: configurable sensitivity (0.0-1.0) and three action modes (redact/warn/block); `leak_detection = false` disables entirely.

## Privacy and Data Hygiene (required)

- Data-hygiene status (`pass|needs-follow-up`): `pass`
- Redaction/anonymization notes: Detected credential patterns are logged at `warn` level with pattern names only (not actual secret values). Redacted content replaces matched strings with `[REDACTED]` before reaching channel adapters or persistent storage. No new PII collection or external reporting.
- Neutral wording confirmation (use ZeroClaw/project-native labels if identity-like wording is needed): Confirmed -- all user-facing messages use neutral, project-native terminology.

## Compatibility / Migration

- Backward compatible? (`Yes/No`): Yes
- Config/env changes? (`Yes/No`): Yes -- new optional `[security.output_guardrail]` TOML section. Defaults to enabled/redact/0.7 sensitivity. Existing configs work unchanged.
- Migration needed? (`Yes/No`): No
- If yes, exact upgrade steps: N/A

## i18n Follow-Through (required when docs or user-facing wording changes)

- i18n follow-through triggered? (`Yes/No`): No
- If `Yes`, locale navigation parity updated in `README*`, `docs/README*`, and `docs/SUMMARY.md` for supported locales (`en`, `zh-CN`, `ja`, `ru`, `fr`, `vi`)? (`Yes/No`): N/A
- If `Yes`, localized runtime-contract docs updated where equivalents exist (minimum for `fr`/`vi`: `commands-reference`, `config-reference`, `troubleshooting`)? (`Yes/No/N.A.`): N/A
- If `Yes`, Vietnamese canonical docs under `docs/i18n/vi/**` synced and compatibility shims under `docs/*.vi.md` validated? (`Yes/No/N.A.`): N/A
- If any `No`/`N.A.`, link follow-up issue/PR and explain scope decision: No user-facing docs or wording changes in this PR.

## Human Verification (required)

What was personally validated beyond CI:

- Verified scenarios: All four output paths tested -- channel adapters via `sanitize_channel_response`, WebSocket via `finalize_ws_response` (including fallback path), CLI interactive REPL, and CLI single-shot `--message` mode.
- Edge cases checked: Empty LLM response with tool output containing credentials (WS fallback path), clean content passthrough, disabled guardrail mode, NaN/out-of-range sensitivity validation, multiple credential patterns in single response.
- What was not verified: Draft streaming path (deferred to Phase 2), production load testing, all 15+ channel adapters end-to-end (tested via shared `sanitize_channel_response` convergence point).

## Side Effects / Blast Radius (required)

- Affected subsystems/workflows: `channels/` (all adapters via shared sanitize function), `gateway/` (HTTP + WebSocket handlers), `agent/` (CLI output loop), `config/` (new config section + validation), `security/` (import cleanup only).
- Potential unintended effects: False positive redaction on strings resembling API key patterns in legitimate LLM responses. Mitigated by configurable sensitivity and `leak_detection = false` escape hatch.
- Guardrails/monitoring for early detection: `tracing::warn!` emitted on every detection with pattern names for log monitoring. Config validation rejects invalid sensitivity values at startup. Three action modes allow gradual rollout (warn - redact - block).

## Agent Collaboration Notes (recommended)

- Agent tools used (if any): Claude Code (research, implementation, review response, rebase conflict resolution)
- Workflow/plan summary (if any): Scout - Research - Design - Implement - CodeRabbit review - Fix - Rebase - PR compliance
- Verification focus: All four output data paths covered, WS fallback bypass fix verified, sensitivity validation boundary conditions
- Confirmation: naming + architecture boundaries followed (`AGENTS.md` + `CONTRIBUTING.md`): Yes

## Rollback Plan (required)

- Fast rollback command/path: Set `leak_detection = false` in `[security.output_guardrail]` config section for immediate passthrough. Or `git revert <merge-commit>` -- all changes are additive with no schema migrations.
- Feature flags or config toggles (if any): `leak_detection` (bool, default true), `leak_action` (redact/warn/block, default redact), `leak_sensitivity` (0.0-1.0, default 0.7)
- Observable failure symptoms: Unexpected `[REDACTED]` strings in agent responses (false positive), or `warn`-level log lines `output guardrail: credential leak detected in outbound message` with pattern names.

## Risks and Mitigations

- Risk: False positive redaction on legitimate content resembling API key patterns
  - Mitigation: Configurable sensitivity (0.7 default), three action modes, `leak_detection = false` kill switch
- Risk: Performance overhead from regex scanning on every response
  - Mitigation: `LeakDetector` uses pre-compiled regex set; scanning is O(n) on response length; typical responses are <10KB

## Test plan

- [x] Unit tests for redact/warn/block modes (`channels::tests`)
- [x] Unit test for disabled guardrail passthrough
- [x] Unit test for clean content (no false positive)
- [x] Integration test: `sanitize_channel_response` with embedded AWS key
- [x] Config deserialization tests (defaults, custom TOML, empty TOML)
- [x] Config validation tests (`leak_sensitivity` out-of-range, NaN, boundary)
- [x] WebSocket fallback path sanitization tests
- [x] `cargo fmt` passes
- [ ] `cargo clippy` (blocked by pre-existing upstream `Cargo.lock` desync)
- [ ] `cargo test --locked` (blocked by pre-existing upstream `Cargo.lock` desync)

## Summary by CodeRabbit

* **New Features**
  * Automatic output guardrails that detect and handle credential leaks (default: enabled, sensitivity 0.7, default action: redact). Applied across CLI (single-shot and interactive), WebSocket, and API responses (including streaming fallbacks).
  * Configurable actions: redact, warn, or block; sensitivity validated to be between 0.0 and 1.0.
* **Tests**
  * Expanded tests covering guardrail behaviors, actions, and sensitivity edge cases.

### CODE REVIEW

File: src/agent/loop_.rs [hunk 1]
```diff
@@ -3168,6 +3168,11 @@ pub async fn run(
                     continue;
                 }
             };
+            // Apply output guardrail (credential leak detection) before sending CLI response.
+            let response = crate::channels::apply_output_guardrail(
+                &response,
+                &config.security.output_guardrail,
+            );
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/agent/loop_.rs:3178]: Fixed in 7332b6db — added `apply_output_guardrail()` to the single-shot CLI path (`--message` branch, line 3042) before `println!`.

File: src/channels/mod.rs [hunk 2]
```diff
@@ -1861,8 +1903,11 @@ async fn process_channel_message(
                 }
             }
 
-            let sanitized_response =
-                sanitize_channel_response(&outbound_response, ctx.tools_registry.as_ref());
+            let sanitized_response = sanitize_channel_response(
+                &outbound_response,
+                ctx.tools_registry.as_ref(),
+                &ctx.output_guardrail,
+            );
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/channels/mod.rs:3116]: Acknowledged — draft streaming sanitization is a valid concern but requires deeper refactoring of the `update_draft` flow across all channel adapters. Tracking this for Phase 2. The current PR covers all *final* output paths.

File: src/channels/mod.rs [hunk 3]
```diff
@@ -3323,6 +3368,7 @@ pub async fn start_channels(config: Config) -> Result<()> {
             None
         },
         non_cli_excluded_tools: Arc::new(config.autonomy.non_cli_excluded_tools.clone()),
+        output_guardrail: config.security.output_guardrail.clone(),
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/channels/mod.rs:4717]: Good point on documentation. Will add a docs section for `[security.output_guardrail]` config reference as a follow-up (or in Phase 2 PR alongside draft streaming). Not blocking for Phase 1 code merge.

File: src/config/schema.rs [hunk 4]
```diff
@@ -3260,6 +3260,57 @@ pub struct SecurityConfig {
...
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct OutputGuardrailConfig {
+    /// Enable credential leak detection in outbound messages. Default: `true`.
+    #[serde(default = "default_true")]
+    pub leak_detection: bool,
+
+    /// Leak detection sensitivity (0.0–1.0, higher = more aggressive).
+    /// Default: `0.7`. Reduce to avoid false positives on technical content.
+    #[serde(default = "default_leak_sensitivity")]
+    pub leak_sensitivity: f64,
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/config/schema.rs:4161]: Fixed in 7332b6db — added bounds validation in `Config::validate()`: rejects NaN, negative, and >1.0 values. Unit test covers out-of-range and boundary values (0.0, 1.0).

File: src/channels/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/channels/mod.rs:3116]: Thanks for the offer\! I'll capture it myself — the draft streaming guardrail and config docs will be part of Phase 2 scope. No need for a separate issue at this point since it's already tracked in the PR description and #1979 phased plan.

File: src/channels/mod.rs [hunk 5]
```diff
@@ -251,6 +251,7 @@ struct ChannelRuntimeContext {
     query_classification: crate::config::QueryClassificationConfig,
     model_routes: Vec<crate::config::ModelRouteConfig>,
     approval_manager: Arc<ApprovalManager>,
+    output_guardrail: crate::config::OutputGuardrailConfig,
 }
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/channels/mod.rs:255]: Fixed in 9716167 — added `output_guardrail: crate::config::OutputGuardrailConfig::default()` to all 14 missing `ChannelRuntimeContext` test literals. Verified with automated analysis: 38/38 instances now include the field, zero missing.

File: src/gateway/mod.rs [hunk 6]
```diff
@@ -965,7 +965,11 @@ async fn run_gateway_chat_with_tools(state: &AppState, message: &str) -> anyhow:
 }
 
 fn sanitize_gateway_response(response: &str, tools: &[Box<dyn Tool>]) -> String {
-    let sanitized = crate::channels::sanitize_channel_response(response, tools);
+    let sanitized = crate::channels::sanitize_channel_response(
+        response,
+        tools,
+        &crate::config::OutputGuardrailConfig::default(),
+    );
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/gateway/mod.rs]: Fixed in 9716167 — `sanitize_gateway_response` now accepts `&OutputGuardrailConfig` as a third parameter. All 7 production call sites extract the runtime config via `&state.config.lock().security.output_guardrail` instead of using `default()`. Test call sites use `OutputGuardrailConfig::default()` explicitly (appropriate for test isolation).

File: src/gateway/ws.rs [hunk 7]
```diff
@@ -95,14 +99,18 @@ fn finalize_ws_response(
...
 
     if let Some(tool_output) = extract_latest_tool_output(history) {
-        let excerpt = crate::util::truncate_with_ellipsis(tool_output.trim(), 1200);
+        // Apply output guardrail to tool output — it may contain leaked credentials.
+        let safe_output =
+            crate::security::apply_output_guardrail(tool_output.trim(), output_guardrail);
+        let excerpt = crate::util::truncate_with_ellipsis(&safe_output, 1200);
         return format!(
             "Tool execution completed, but the model returned no final text response.\n\nLatest tool output:\n{excerpt}"
         );
```
[the remark this answers is not in the document]
Comment (Mike-7777777) [reply src/gateway/ws.rs:124]: Fixed in 12d00ed -  `finalize_ws_response` now detects Block mode via `matches!(config.leak_action, LeakAction::Block) && safe_output != trimmed` and returns the policy message directly without wrapping it as "Latest tool output:".

Design note: CodeRabbit's suggested `safe_output.trim().is_empty()` check wouldn't trigger because Block mode returns a non-empty warning string, not an empty string. Instead, this fix uses config + input/output comparison - no magic string matching, type-safe and maintainable.

Two boundary tests added:
- `finalize_ws_response_returns_block_message_directly_when_tool_output_blocked`
- `finalize_ws_response_shows_tool_output_normally_when_block_mode_but_clean_content`

**Comment by Mike-7777777** [comment]
**Re: CodeRabbit round-4 review on `12d00ed`**

**1. "Fix rustfmt drift" (Outside diff, Minor)**
False positive — CI formatting failure is caused by upstream `Cargo.lock` desync (`--locked` flag), not by rustfmt drift. `cargo fmt --all -- --check` passes clean locally. Upstream `main` CI has the same `Cargo.lock` issue. No action needed.

**2. "Add rollout/rollback note near wiring" (Nitpick)**
Already covered in the PR description under **Rollback Plan** section: `leak_detection = false` kill switch, `leak_action` toggle (redact/warn/block), and `git revert` path. Adding duplicate operator notes in code comments would violate DRY. Configuration reference docs are tracked for Phase 2.

**Comment by theonlyhennygod** [comment]
Superseded by #2007, which landed the focused LeakDetector outbound sanitization fix and closed #1979.
